### PR TITLE
CosmosDB Periodic Backups and Retention

### PR DESCRIPTION
**Pull request by troy0820**
### Which issue this PR addresses:

Fixes [10196483](https://msazure.visualstudio.com/AzureRedHatOpenShift/_workitems/edit/10196483/)

### What this PR does / why we need it:

This PR allows us to update the cosmosdb `azure-sdk-for-go` package to allow for cosmosdb databases to be backed up periodically and the retention to be set explicitly.  The retention is 30 days or 720 hours and the backup intervals are 4 hours or 240 minutes.  

- [x] Add package `documentdb` version `2021-01-15`
- [x] Add backup interval for 240 mins and retention for 720 hours _(4 hours and 30 days respectively)_
- [x] Add `ipRules` to `rp-production-parameters.json` _(necessary to generate arm template with existing data)_
- [x] Removed `extraCosmosDBIPs` from `rp-production-parameters` _(still in use from deployer but not an active param for arm template)_
### Test plan for issue:

Existing unit/e2e tests and manual deploy to INT environment test arm template deployment. 
- [x] ~[INT deployment results](https://msazure.visualstudio.com/AzureRedHatOpenShift/_build/results?buildId=44311912&view=logs&j=92338bc8-46cf-53af-189a-75616d32f690)~
- [x] ~[INT deployment after rebase results](https://msazure.visualstudio.com/AzureRedHatOpenShift/_build/results?buildId=44337848&view=logs&j=ac5ecb5a-030b-5f61-b782-cded586d007c&t=ac5ecb5a-030b-5f61-b782-cded586d007c)~
- [x] [INT deployment results](https://msazure.visualstudio.com/AzureRedHatOpenShift/_build/results?buildId=45075143&view=results) 
### Is there any documentation that needs to be updated for this PR?

No. Documentation for the update to the new sdk-for-go is not necessary as the new capability is built in the configuration.

**NOTE: This does not remove `extraCosmosDBIPs` from RP-Config but uses those values (strings) to dynamically allocate the specific data type necessary for `ipRules` without changing anything.**

### CODE REVIEW

File: deploy/rp-production.json [hunk 1]
```diff
@@ -712,10 +712,17 @@
                     }
                 ],
                 "databaseAccountOfferType": "Standard",
-                "ipRangeFilter": "[if(parameters('disableCosmosDBFirewall'), '', concat('104.42.195.92,40.76.54.131,52.176.6.30,52.169.50.45,52.187.184.26', if(equals(parameters('extraCosmosDBIPs'), ''), '', ','), parameters('extraCosmosDBIPs')))]",
+                "ipRules": "[concat(if(parameters('disableCosmosDBFirewall'), createArray(), parameters('disableCosmosDBFirewall')), if(parameters('extraCosmosDBIPs'),parameters('extraCosmosDBIPs'), createArray())]",
```
Comment (troy0820) [inline deploy/rp-production.json:715]: This is dependent on us changing the structure of the data that is presently a string pointer.  The type needs to be a []ipRangeOrAddress which we can do in the "configuration" to allow this to work.

Comment (mjudeikis) [reply deploy/rp-production.json:715]: How do you think this changed in config repository? 
Currently we have "global hardcoded" IPs and per region ones. So overrides dont quite works. 
Can you give an example of how config will change?

Comment (troy0820) [reply deploy/rp-production.json:715]: >How do you think this changed in config repository?

I don't believe they changed, but the newer version of the documentdb forces us to change the structure of the data in the config as we can not dynamically create an array of type `ipAddressOrRange` for each IP address where before they were string pointers being constructed as a string split by commas.

Currently, the config shows the IPs as "strings" but making them an array of type `ipAddressOrRange` should suffice.  
```yaml
extraCosmosDBIPs:
- ipAddressOrRange: some-ip-we-have
- ipAddressOrRange: some-ip-we-have
- ipAddressOrRange: some-ip-we-have
```

File: pkg/deploy/generator/resources_rp.go [hunk 2]
```diff
@@ -1311,9 +1318,8 @@ func (g *generator) rpCosmosDB() []*arm.Resource {
 		Resource:   cosmosdb,
 		APIVersion: azureclient.APIVersion("Microsoft.DocumentDB"),
 	}
-
 	if g.production {
-		cosmosdb.IPRangeFilter = to.StringPtr("[if(parameters('disableCosmosDBFirewall'), '', concat('104.42.195.92,40.76.54.131,52.176.6.30,52.169.50.45,52.187.184.26', if(equals(parameters('extraCosmosDBIPs'), ''), '', ','), parameters('extraCosmosDBIPs')))]")
```
Comment (mjudeikis) [inline pkg/deploy/generator/resources_rp.go:1316]: We are dropping these on the floor. 
We have few types of inputs here:
1. Standalone hardocoded public IPS: https://docs.microsoft.com/en-us/azure/cosmos-db/how-to-configure-firewall#allow-requests-from-the-azure-portal 
2. And private ones we provide for AME and billing.

Currently we do merge configs in our flow where if override is not provided we take one from Configuration stanza.
We need to make sure those are not gone.

They could be added into config, but we don't have "union" feature in our config merger functionality. 

With the approach you are doing now I dont from top of my head how to achieve result we need (IP from config + wellknown values) as a parameter.

Sorry for gibrish here but I think we need to step back and see how we can do this better. Maybe even changing way we construct these ARM files. :exploding_head:

Comment (troy0820) [reply pkg/deploy/generator/resources_rp.go:1316]: No, it makes sense. I agree that these values get lost if the value for `disableCosmosDBFIrewall` is not set and these values should _not_ be hardcoded and added to the configuration. 

The challenge here is to keep these values while allowing the other values from the template parameters to be added to the array from the configuration. 

A wider discussion may be necessary for how we do arm templates going forward.

Comment (mjudeikis) [reply pkg/deploy/generator/resources_rp.go:1316]: Yes, but we can't just copy paste save list in the config for all 30+ regions. 
This is why config merging might need to change. Currently this PR can't be merged due to this :/

File: deploy/rp-production.json (hunk 1 above)
Comment (troy0820) [reply deploy/rp-production.json]: Data is not generated but overridden in the `pkg/deploy/deploy_rp.go` where `mgmtdocumentdb.IPAddressOrRange` array is populated by the `extraCosmosDBIPs` from configuration as well as `disableCosmosDBFIrewall` parameters.

File: pkg/deploy/generator/resources_rp.go (hunk 2 above)
Comment (troy0820) [reply pkg/deploy/generator/resources_rp.go:1316]: As discussed: Data is not generated but overridden in the pkg/deploy/deploy_rp.go where mgmtdocumentdb.IPAddressOrRange array is populated by the extraCosmosDBIPs from configuration as well as disableCosmosDBFIrewall parameters.

**Comment by troy0820** [comment]
Deploying to INT is necessary to see if changes populated through `pkg/deploy/deploy_rp.go` consistently capture the necessary configuration from the overridden parameters.

**Comment by nilsanderselde** [comment]
The recent changes look good to me

**Comment by troy0820** [comment]
> The recent changes look good to me

Deploying to INT to see if the changes work with the "configuration", results can be posted to verify it works and WIP will be taken out of draft.

**Review by m1kola — DISMISSED** [review]
Test, ignore

**Comment by troy0820** [comment]
Successful restore of cosmosdb database achieved in INT subscription.

#### NOTE:  This is not a restore PR but changes the frequency of how we rely on backups and the intervals it takes to create them.

**Review by jewzaam — DISMISSED** [review]
I'd like INT deployment including billing E2E green after this merges.  Today sounds like a good time to do the merge and get it tested, check the tests Friday.

**Comment by troy0820** [comment]
@jewzaam Tests are green from the INT deployment with this branch and images pushed to INT.  
- [x] [INT deployment results](https://msazure.visualstudio.com/AzureRedHatOpenShift/_build/results?buildId=45075143&view=results)

**Review by jewzaam — APPROVED** [review]

**Pull request merged**